sycl/include/sycl/ext/oneapi/experimental/invoke_simd.hpp

```diff
@@ -331,6 +345,42 @@ template <typename T>
 using remove_ref_from_func_ptr_ref_type_t =
     typename remove_ref_from_func_ptr_ref_type<T>::type;
 
+template <typename T> struct strip_regcall_from_function_ptr;
+
+template <typename Ret, typename... Args>
+struct strip_regcall_from_function_ptr<Ret (*)(Args...)> {
+  using type = Ret (*)(Args...);
+};
+
+template <typename Ret, typename... Args>
+struct strip_regcall_from_function_ptr<Ret(__regcall *)(Args...)> {
+  using type = Ret (*)(Args...);
+};
+
+template <typename T>
+using strip_regcall_from_function_ptr_t =
+    typename strip_regcall_from_function_ptr<T>::type;
+
+template <typename Ret, typename... Args>
+constexpr bool has_ref_arg(Ret (*)(Args...)) {
+  return std::is_reference_v<Ret> || (... || std::is_reference_v<Args>);
+}
+
+template <class Callable> constexpr void verify_no_ref() {
+  if constexpr (is_function_ptr_or_ref_v<Callable>) {
+    using RemoveRef =
+        remove_ref_from_func_ptr_ref_type_t<std::remove_reference_t<Callable>>;
+    using FuncPtrType =
+        std::conditional_t<std::is_pointer_v<RemoveRef>, RemoveRef,
+                           std::add_pointer_t<RemoveRef>>;
+    using FuncPtrNoCC = strip_regcall_from_function_ptr_t<FuncPtrType>;
+    constexpr FuncPtrNoCC obj = {};
+    constexpr bool callable_has_ref_arg_or_ret = has_ref_arg(obj);
+    static_assert(!callable_has_ref_arg_or_ret,
+                  "invoke_simd does not support references");
+  }
```

Comment:
If split the func above, then the message can be even more specific: not ret or arg, but tell exactly either in ret or in arg.
```suggestion
    constexpr bool callable_has_ref_ret = has_ref_ret(obj);
    static_assert(!callable_has_ref_ret,
                  "invoke_simd does not support callee returning a reference");
    constexpr bool callable_has_ref_arg = has_ref_arg(obj);
    static_assert(!callable_has_ref_arg,
                  "invoke_simd does not support callee with reference parameters");
  }
```